### ISSUE TITLE
FileContextMode::Node is unsound 

### ISSUE THREAD

**Issue by rydrman** [SnowflakePowered/winfsp-rs#20]
Hello! I am trying to get my feet wet with this library, and so trying to create a filesystem that simply presents an empty folder. I am able to get it running after implementing `get_security_by_name` for the root directory, and a `read_directory` implementation that returns `Ok(0)` but trying to enter the directory in windows explorer yields:

```
process didn't exit successfully: `target\debug\my-winfsp.exe -` (exit code: 0xc0000374, STATUS_HEAP_CORRUPTION)
```

It's not clear if this is a problem with my lack of windows knowledge or something up with the bindings themselves, any help would be appreciated!

**Comment by chyyran** [comment]
Hi, could you send me some example code to investigate? STATUS_HEAP_CORRUPTION should not be happening in safe code and is likely a library bug.

**Comment by rydrman** [comment]
Thanks for the quick reply! I've boiled down my code into something small, trying to keep the gist of everything that I was doing -

https://github.com/rydrman/winfsp-minimal-repro

**Comment by chyyran** [comment]
Narrowed this down to a double free due to WinFSP calling `close` twice.

I guess this is the expected behaviour for `FileContextMode::Node`. For the time being, you should use `FileContextMode::Descriptor` and that should prevent the double-free. Its also consistent with your usage, since you’re returning a new instance of `FileContext` on each open.

That doesn’t bode well for the Node mode, but I think it should be able to be made sound by implicitly wrapping everything in an Arc when opening a file in the Node mode. I’ll be looking into that soon.

**Comment by chyyran** [comment]
After a brief investigation it doesn’t seem the current design works well for FileContextMode::Node.

> When both of these flags are unset (default), the FileContext parameter represents the file node. The file node is a void pointer (or an integer that can fit in a pointer) that is used to uniquely identify an open file. Opening the same file name should always yield the same file node value for as long as the file with that name remains open anywhere in the system.


Currently the design of the `FileSystemContext` trait encouraged the descriptor pattern. I’ve managed to find a way to make `Node` sound but in turn its utility at least in Rust seems to be less useful than initially thought.

**Comment by chyyran** [comment]
On further investigation I've concluded that there's no possible way `FileContextMode;:Node` can be made sound. Most of its benefits can be replicated in Rust by returning `Arc` if wanted.

**Comment by chyyran** [comment]
I've fixed this in #21 by removing `FileContextMode` completely. I think the internal logic of the crate makes too many assumptions for `FileContextMode::Node` to be sound without a complete rewrite, and frankly I don't think it's possible to write a sound (in Rust) wrapper with the assumptions that `FileContextMode::Node` implies. 

It's not too much lost, since equivalent logic can be provided by the file context by just doing bookkeeping themselves with `Arc`. Note that the reference [memfs](https://github.com/winfsp/winfsp/blob/master/tst/memfs/memfs.cpp#L449) effectively does the same.

**Comment by rydrman** [comment]
Thanks again - as far as I've seen your decision makes sense to me!
